### ISSUE TITLE
Proper solution to figure out terminal size inside of a subshell

### ISSUE THREAD

**Issue by purpleP** [crossterm-rs/crossterm#419]
My recent [investigation](https://users.rust-lang.org/t/why-does-rust-binary-take-so-much-space/41088/12) about my binary size have led me to the fact that crossterm wastes 10K to call tput to determine terminal size in a subshell.

I've found the [discussion](https://github.com/crossterm-rs/crossterm/issues/276).

It hurts my feelings that such things are merged into master especially without even tracking issue to make proper solution.

I've created a [question](https://stackoverflow.com/questions/61288198/how-to-debug-tput) to debug tput inside of a subshell.

@Canop when running inside gdb it should count as a subshell, right?

```
--with-install-prefix=/tmp/delme/tmp/delme/bin/tputSYS_ioctl(2, 0x5401, 0x7ffda84b6480, 0x55ec6d0d4010)             = 0
SYS_ioctl(2, 0x5401, 0x7ffda84b64b0, 0x55ec6d0d4010)             = 0
SYS_ioctl(2, 0x5401, 0x7ffda84b64b0, 0x55ec6d0d4010)             = 0
SYS_ioctl(2, 0x5401, 0x7ffda84b6410, 0x55ec6d0d4010)             = 0
SYS_ioctl(2, 0x5413, 0x7ffda84b64d0, 0x55ec6d0d4010)             = 0
env TERM=xterm-kitty gdbserver localhost:2345 progs/tput cols(ins)(gdb) b tput_cmd
Breakpoint 1 at 0x555555556dd0
(ins)(gdb) c
Continuing.
Reading /lib/x86_64-linux-gnu/libc.so.6 from remote target...
Reading /lib/x86_64-linux-gnu/libc-2.30.so from remote target...
Reading /lib/x86_64-linux-gnu/.debug/libc-2.30.so from remote target...
Reading /usr/lib/debug//lib/x86_64-linux-gnu/libc-2.30.so from remote target...
Reading /usr/lib/debug//lib/x86_64-linux-gnu/libc-2.30.so from remote target...

Breakpoint 1, 0x0000555555556dd0 in tput_cmd ()
(ins)(gdb) l
1       ../sysdeps/unix/sysv/linux/x86_64/init-first.c: No such file or directory.
(ins)(gdb) l
1       in ../sysdeps/unix/sysv/linux/x86_64/init-first.c
(ins)(gdb) l
1       in ../sysdeps/unix/sysv/linux/x86_64/init-first.c
(ins)(gdb) n
Single stepping until exit from function tput_cmd,
which has no line number information.
0x00005555555569fe in main ()
(ins)(gdb)
ltrace -S bash -c 'tput cols' 2>&1 | tail -n20
SYS_stat("/etc/terminfo", 0x562b5cb264d0)        = 0
SYS_stat("/lib/terminfo", 0x562b5cb26560)        = 0
SYS_stat("/usr/share/terminfo", 0x562b5cb265f0)  = 0
SYS_access("/etc/terminfo/x/xterm-256color", 04) = -2
SYS_access("/lib/terminfo/x/xterm-256color", 04) = 0
SYS_openat(0xffffff9c, 0x7ffd9a2a57a0, 0, 0)     = 3
SYS_fstat(3, 0x7ffd9a295530)                     = 0
SYS_read(3, "\036\002%", 32768)                  = 3503
SYS_read(3, "", 28672)                           = 0
SYS_close(3)                                     = 0
SYS_ioctl(0, 0x5401, 0x7ffd9a2a6720, 0x562b5cb26010) = 0
SYS_ioctl(0, 0x5401, 0x7ffd9a2a6750, 0x562b5cb26010) = 0
SYS_ioctl(0, 0x5401, 0x7ffd9a2a6750, 0x562b5cb26010) = 0
SYS_ioctl(0, 0x5401, 0x7ffd9a2a66b0, 0x562b5cb26010) = 0
SYS_ioctl(0, 0x5413, 0x7ffd9a2a6770, 0x562b5cb26010) = 0
SYS_fstat(1, 0x7ffd9a2a5ef0)                     = 0
SYS_write(1, "106\n", 4106
)                         = 4
SYS_exit_group(0 <no return ...>
+++ exited (status 0) +++
```

It seems like `tput` isn't doing anything except invoking `ioctl`.

Have anyone built minimal reproducible example using ioctl which gives incorrect answer in a subshell? Is there a test that shows that it doesn't work?

**Comment by purpleP** [comment]
I have no idea what I'm doing here but mere bruteforce and googling resulted in this
```
fn main() {
    let mut size = winsize {
        ws_row: 0,
        ws_col: 0,
        ws_xpixel: 0,
        ws_ypixel: 0,
    };
    let fd = std::fs::OpenOptions::new()
        .read(true)
        .write(true)
        .open("/dev/tty")
        .map(|file| file.as_raw_fd())
        .unwrap()
        ;
    unsafe { ioctl(fd, TIOCGWINSZ.into(), &mut size) };
    println!("/dev/tty cols: {}, lines: {}", size.ws_row, size.ws_col);
    unsafe { ioctl(STDOUT_FILENO, TIOCGWINSZ.into(), &mut size) };
    println!("stdout cols: {}, lines: {}", size.ws_row, size.ws_col);
    unsafe { ioctl(STDERR_FILENO, TIOCGWINSZ.into(), &mut size) };
    println!("stderr cols: {}, lines: {}", size.ws_row, size.ws_col);
    unsafe { ioctl(STDIN_FILENO, TIOCGWINSZ.into(), &mut size) };
    println!("stdin cols: {}, lines: {}", size.ws_row, size.ws_col);
}

```

```
echo (cargo run 2>&1)
   Compiling test_term_size v0.1.0 (/home/m/code/test_term_size) warning: unused import: `File`  --> src/main.rs:2:15   | 2 | use std::fs::{File};   |               ^^^^   |   = note: `#[warn(unused_imports)]` on by default  warning: unused imports: `O_RDWR`, `open`  --> src/main.rs:4:5   | 4 |     O_RDWR,   |     ^^^^^^ 5 |     open, ioctl, winsize, STDIN_FILENO,   |     ^^^^      Finished dev [unoptimized + debuginfo] target(s) in 0.21s      Running `target/debug/test_term_size` /dev/tty cols: 0, lines: 0 stdout cols: 0, lines: 0 stderr cols: 0, lines: 0 stdin cols: 52, lines: 106
```
@Canop  does that solves the problem in your usecase?

**Comment by purpleP** [comment]
Few other comments about code quality.

You could've at least use one subprocess instead of two.
```
tput -S <<!
(ins)> cols
(ins)> lines
(ins)> !
106
52
```

Where's `Ok(false)` and if it's not there then why isn't it is `Ok(())`?
``` 
pub fn wrap_with_result(result: i32) -> Result<bool> {
    if result == -1 {
        Err(ErrorKind::IoError(io::Error::last_os_error()))
    } else {
        Ok(true)
    }
}
```

**Comment by TimonPost** [comment]
I am not familiar with TPUT, that time I was still developing on Windows. Most of the ideas for TPUT were discussed in the discord server and not reported back to that issue. @Canop needed the feature and thus created a PR. We could not find any reasons why not to use tput, with the only downside being that it was a subprocess.

> It hurts my feelings that such things are merged into master especially without even tracking issues to make a proper solution. & You could've at least use one subprocess instead of two.

If you have a solution or improvement feel free to create an issue for it or open a PR. It is how it is and we all try to do the best here. Complaining about feelings and ranting (or at least, that is what it sounds like) about code quality isn't going to fix anything.  

> I've created a question to debug tput inside of a subshell.

The question does not exist anymore

> Where's Ok(false) and if it's not there then why isn't it is Ok(())?

I am not sure if I can follow your question, can you rephrase it?

You have not defined what your problem is. 

-  What is your issue with TPUT?
-  What is the problem that you see?
-  And does your problem occure with GDC?

**Comment by purpleP** [comment]
> I am not familiar with TPUT, that time I was still developing on Windows. Most of the ideas for TPUT were discussed in the discord server and not reported back to that issue. @Canop needed the feature and thus created a PR. We could not find any reasons why not to use tput, with the only downside being that it was a subprocess.
> 
> > It hurts my feelings that such things are merged into master especially without even tracking issues to make a proper solution. & You could've at least use one subprocess instead of two.
> 
> If you have a solution or improvement feel free to create an issue for it or open a PR. It is how it is and we all try to do the best here. Complaining about feelings and ranting (or at least, that is what it sounds like) about code quality isn't going to fix anything.
> 
> > I've created a question to debug tput inside of a subshell.
> 
> The question does not exist anymore
> 
> > Where's Ok(false) and if it's not there then why isn't it is Ok(())?
> 
> I am not sure if I can follow your question, can you rephrase it?
> 
> You have not defined what your problem is.
> 
>     * What is your issue with TPUT?
> 
>     * What is the problem that you see?
> 
>     * And does your problem occur 'only; with GDC?

1.

**Comment by purpleP** [comment]
@TimonPost 

I'm not sure what to answer to have a productive discussion and not hurt anyone's feelings, but I'll try my best (I was thinking how to answer for at least an hour).

> with the only downside being that it was a subprocess.

If wasting cycles on calling a subprocess twice (that's two system calls) to call a program that would issue a few system calls itself to do the thing that should've been accomplished by one system call seems okay to you.

If not understanding the domain you working with to the point that you're okay to call another program to do what you should be doing yourself **hoping** it would do the right thing seems okay to you.

If **implicitly depending on ncurses** without admitting it in README in a big bold font seems okay to you.

then our code quality standards differ enough that further discussion whether this was right decision or wrong one is completely pointless.

Now a bit about possible solution.

```
use std::os::unix::io::AsRawFd;
use libc::{
    ioctl, winsize, STDIN_FILENO,
    STDOUT_FILENO,
    STDERR_FILENO,
    TIOCGWINSZ,
};


fn main() {
    let mut size = winsize {
        ws_row: 0,
        ws_col: 0,
        ws_xpixel: 0,
        ws_ypixel: 0,
    };
    let file = std::fs::OpenOptions::new()
        .read(true)
        .write(true)
        .open("/dev/tty")
        .unwrap()
        ;
    unsafe { ioctl(file.as_raw_fd(), TIOCGWINSZ.into(), &mut size) };
    println!("/dev/tty cols: {}, lines: {}", size.ws_row, size.ws_col);
    unsafe { ioctl(STDOUT_FILENO, TIOCGWINSZ.into(), &mut size) };
    println!("stdout cols: {}, lines: {}", size.ws_row, size.ws_col);
    unsafe { ioctl(STDERR_FILENO, TIOCGWINSZ.into(), &mut size) };
    println!("stderr cols: {}, lines: {}", size.ws_row, size.ws_col);
    unsafe { ioctl(STDIN_FILENO, TIOCGWINSZ.into(), &mut size) };
    println!("stdin cols: {}, lines: {}", size.ws_row, size.ws_col);
}
```
This gives

```
echo (cargo run 2>&1)
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s      Running `target/debug/test_term_size` /dev/tty cols: 25, lines: 213 stdout cols: 25, lines: 213 stderr cols: 25, lines: 213 stdin cols: 25, lines: 213
```

Judging by [man](http://man7.org/linux/man-pages/man4/tty.4.html) and answer on [stackoverflow](https://stackoverflow.com/questions/61294762/which-file-descriptor-should-be-used-in-ioctl-to-know-terminal-screen-size?noredirect=1#comment108439008_61294762) on linux you should use `/dev/tty` as terminal file descriptor, not stdout or stderr, nor stdin.

> If you have a solution or improvement feel free to create an issue for it or open a PR

You see, It's completely true that you don't owe me anything. But when you're answering "if you don't like anything open PR" you should understand that for me to spend time on that I need to be sure it's not a waste of time. If you will have same policies on including such solutions without at least making sure they temporary (tracking issues, deadlines on removing workarounds etc) contributing to this project will certainly be waste of time for me.

I see that you already have a pull request that introduces `tty_fd` IO action so I think there's no point creating merge conflicts by me creating new pull request.

Now
>> Where's Ok(false) and if it's not there then why isn't it is Ok(())?

> I am not sure if I can follow your question, can you rephrase it?

The question is what is the purpose of the `bool` inside `Result<bool`>?.

**Comment by purpleP** [comment]
By the way I've modified code to use `dev/tty` and my binary is smaller by 28K. I'm rewriting less pager and less is ~160K (also crossplatform by the way) so shaving off 28K is no small difference.

**Comment by sharnoff** [comment]
Hey @purpleP.

This comment is split into two - the second is given as a separate comment immediately following this one, as the two pertain to entirely different subjects. The first is a response to your communication style and the second addresses the issue with ioctl.

### A response to your communication style

Who am I? I'm some guy that opened a couple PRs at some point in this repo - I watch it because I'm using crossterm pretty heavily for a project and I want to generally get a sense of what's going on. I'm writing up this reply to you because it pains me to see the way that you've treated someone who has made their work availble to the public for free - **this type of behavior is why people decide to stop giving back to the open-source community**.

You say that @TimonPost doesn't owe you anything. Not only is this true, but it's exactly the way that it shoud be treated; open source collaboration **only works when it is based on a foundation of mutal respect and willingness to help**. All parties involved must approach with clear commuincation, starting with a current problem (ideally providing as much information as possible) or a change they'd like to see made. In both cases - unless you're one of the owners of the repository - productive dialogue **cannot happen** without both parties acknowledging that they must work together to accomplish anything.

To what you said earlier:

> you should understand that for me to spend time on that I need to be sure it's not a waste of time

That's okay; not every project is going to have the things that you want. Perhaps crossterm isn't what you're looking for!

For more information on this topic, feel welcome to have a look at the [GitHub community guidelines](https://help.github.com/en/github/site-policy/github-community-guidelines#building-a-strong-community), which you agreed that you [would not violate "under any circumstances"](https://help.github.com/en/github/site-policy/github-terms-of-service#c-acceptable-use).

**Comment by sharnoff** [comment]
Hopefully this comment can summarize the relevant context from this thread.

I think there are points worth discussing that you bring up - perhaps rephrasing might result in better engagement. I'll mention a few things here, but I'll primarily collect things together so that something might ctually come of this.

The previous discussion you linked has a lot of important context that hasn't been mentioned in this thread yet. Ones that I think are particularly of note include [this comment](https://github.com/crossterm-rs/crossterm/issues/276#issuecomment-545319761) from Canop:

> I agree that invoking a separate program is problematic (should we use it in a separate function?) but you can easily check in a subshell that this solution works while the TIOCGWINSZ ioctl currently doesn't (and doesn't work either in a container).
>
> I'd be happy to use a better solution, using tput is a workaround. Can you investigate and determine why tput works in subshells while the current solution in use in crossterm and similar libs doesn't ?

This comment seems to be what you're responding to, and is followed up by [this example](https://github.com/crossterm-rs/crossterm/issues/276#issuecomment-563899718):

> Write a `sub_program` program needing to know the size of the terminal. Let's imagine it writes that size on stdout.
>
> Then call it as
> ```
> echo "my program says $(sub_program)"
> ```
> Then `sub_program` runs in a subshell and the previous way of obtaining the terminal dimension didn't work.
>
> What tput does could probably be reproduced, and I'm all for doing it, but in the meantime calling it is a working workaround.

To collect everything together, it seems that you tried this with a sample program and it displayed output that *might* be on the path to a solution. After running
```
echo (cargo run 2>&1)
```
with the sample program you gave above, it seems that the result was contrary to what was previously said (though rows an columns seem to be swapped - I fixed that for ya) It seems you may be using a different shell, though - for the sake of completeness I tested your solution in bash  as well, using the executable directly:
```
$ echo "the program outputs: $(./target/debug/test-ground)"
the program outputs: /dev/tty cols: 239, lines: 51
stdout cols: 239, lines: 51
stderr cols: 239, lines: 51
stdin cols: 239, lines: 51
```

<details>
<summary> Updated sample code </summary>

<pre><code>
use libc::{ioctl, winsize, STDERR_FILENO, STDIN_FILENO, STDOUT_FILENO, TIOCGWINSZ};
use std::os::unix::io::AsRawFd;

fn main() {
    let mut size = winsize {
        ws_row: 0,
        ws_col: 0,
        ws_xpixel: 0,
        ws_ypixel: 0,
    };

    let file = std::fs::OpenOptions::new()
        .read(true)
        .write(true)
        .open("/dev/tty")
        .unwrap();

    unsafe { ioctl(file.as_raw_fd(), TIOCGWINSZ.into(), &mut size) };
    println!("/dev/tty cols: {}, lines: {}", size.ws_col, size.ws_row);

    unsafe { ioctl(STDOUT_FILENO, TIOCGWINSZ.into(), &mut size) };
    println!("stdout cols: {}, lines: {}", size.ws_col, size.ws_row);

    unsafe { ioctl(STDERR_FILENO, TIOCGWINSZ.into(), &mut size) };
    println!("stderr cols: {}, lines: {}", size.ws_col, size.ws_row);

    unsafe { ioctl(STDIN_FILENO, TIOCGWINSZ.into(), &mut size) };
    println!("stdin cols: {}, lines: {}", size.ws_col, size.ws_row);
}
</code></pre>
</details>

I don't personally know what other information is required to judge whether this is a change that can be made. @TimonPost, @Canop - do you recall any other test cases that might break this? It seemed from the discussion linked above that the sentiment was to leave this as a temporary solution with the eventual plan of reimplementing `tput` here.

Hopefully this helps - I'd personally love to see a better solution here, but if it isn't a priority, that's completely understandable. I would be willing to try to implement this if it is a workable solution.

**Comment by TimonPost** [comment]
> The question is what is the purpose of the bool inside Result<bool>?.

Ah, so this actually can just be changed to `Ok(())`, probably has to do something with historical reasons. Thanks for pointing out.

> If implicitly depending on ncurses without admitting it in README in a big bold font seems okay to you.

Why do you think that crossterm uses nurses? At least as far I know this isn't the case at all. 

> you should use /dev/tty as terminal file descriptor, not stdout or stderr, nor stdin. &&
I see that you already have a pull request that introduces tty_fd IO action so I think there's no point creating merge conflicts by me creating new pull requests.

Crossterm has been using `/dev/tty` for some time now. However, there were still some places where this wasn't the case. The PR updated the last occurrences where `/dev/tty `was not used yet, however, there was still an error on MacOs which prevents this PR from being merged. Unfortunately, I am not able to test on MacOs. Next to that, this PR is not related to the terminal size issue since it already uses `dev/tty` (see below for more details).

> Call another program to do what you should be doing yourself hoping it would do the right thing seems okay to you.

I can understand that this might be not so good of an idea.  And it would be more than ideal to have this within the library. Your concern about the subprocess was a concern discussed in the linked issue and the PR: #283 as well. It was an temporary solution to a problem that was fixed by TPUT. 

> I don't personally know what other information is required to judge whether this is a change that can be made.

If we can show that we can read the terminal size in both subshell and normal shell with this approach and also validate it's working on a macOS machine we can merge the new implementation and get rid of the TPUT process call. 

> Hopefully this helps - I'd personally love to see a better solution here

I would love to see this improved. I think we can all agree that calling a subprocess is not the best solution. Any improvement which brings this in the house will be welcomed warmly in the source code haha. 

**Current Implementation**

The current implementation tries to use `/dev/tty` in all possible cases where it is possible. If this is not possible it uses `STDOUT_FILENO`. 

```rust
    let file = File::open("/dev/tty").map(|file| (FileDesc::new(file.into_raw_fd(), true)));
    let fd = if let Ok(file) = &file {
        file.raw_fd()
    } else {
        // Fallback to libc::STDOUT_FILENO if /dev/tty is missing
        STDOUT_FILENO
    };
```

Then it will try to use the normal `ioctl` to fetch the terminal size with the handle from above. If this is not possible 'then' it will use TPUT for it. 

```rust
    if let Ok(true) = wrap_with_result(unsafe { ioctl(fd, TIOCGWINSZ.into(), &mut size) }) {
        Ok((size.ws_col, size.ws_row))
    } else {
        tput_size().ok_or_else(|| std::io::Error::last_os_error().into())
    }
```

(https://github.com/crossterm-rs/crossterm/blob/master/src/terminal/sys/unix.rs#L26)

To my understanding, with this approach, we do not call 'tput' for every case only the cases were the 'normal' and 'good' approach does not work. So my question to the proposed question is: how does this improve the scenario?

**Comment by purpleP** [comment]
@TimonPost

> Why do you think that crossterm uses nurses? At least as far I know this isn't the case at all.

Well, where do you think `tput` command is coming from? I don't know details about this curses programs, but one of the implementation of it is definitely comes from `ncurses`. So you kind of made you library a wrapper around ncurses anyway (correct me if I'm wrong) :-) 

> Unfortunately, I am not able to test on MacOS

My work notebook is mac, If you have automatic test that isn't working or you can at least specify steps to reproduce a problem - I could try.

> It was a temporary solution to a problem that was fixed by TPUT

As I've said I understand that you don't own me anything, so I'm just expressing my opinion, right? My experience show that

> there's noting more permanent then temporary

is not a fun pun it's the truth of life in our industry. So I'll repeat it again - unless you've made a process to make sure that temporary solution is indeed temporary and defined "temporary" (is that a half a year, a year, two years, until the death of a universe) you can't call it temporary. Just because of formal logic, you know... As I keep saying to my colleagues - instead of writing **TODO** comments you should write **NEVERDO** because it's more accurate.

**Comment by sharnoff** [comment]
> To my understanding, with this approach, we do not call 'tput' for every case only the cases were the 'normal' and 'good' approach does not work. So my question to the proposed alternative is: how does this improve the current scenario?

That seems to me like a sensible way of doing it... It seems to me that we don't actually know whether 'tput' is ever needed, so ideally we'd be able to to know that (either removing the fallback or figuring out the precise conditions for requiring it). I suspect that for most of the cases where the above attempts fail (e.g. running as a daemon), we would want to return an error anyways, but that's just a guess.

As for the benefits of this - @purpleP: you mentioned binary size as a factor. I wasn't sure from what you wrote above - did you end up testing the reduction in binary size from removing the call to 'tput'? If it was significant, this might be a good reason to make the change - assuming that we don't need 'tput'.

**Comment by purpleP** [comment]
@sharnoff 
> did you end up testing the reduction in binary size from removing the call to 'tput'?

Yes, as I've written 28K decrease in binary size.

> That seems to me like a sensible way of doing it

No comments.

**Comment by sharnoff** [comment]
@PurpleP - What was the 28K decrease out of?

**Comment by purpleP** [comment]
> @purpleP - What was the 28K decrease out of?

From removing `tput` workaround from crossterm.

**Comment by sharnoff** [comment]
Ah, my bad - I meant something different. Let me rephrase: What was the original size of the binary before decreasing?

**Comment by purpleP** [comment]
> Ah, my bad - I meant something different. Let me rephrase: What was the original size of the binary before decreasing?

around 210K, but it doesn't matter. If you're think that it's okay to rely on external command here, we have nothing to discuss, I'm sorry, there's just no point in wasting time on this.

**Comment by TimonPost** [comment]
I am getting a bit tired of the way you are communicating even after this was made clear and we ignored those bad communication styles. I think we all agreed on the fact that the subprocess call is a bad idea however you keep the sticking finger up. We are here to work on a solution, and we do this all in our free time, it is not only your precious time. I will not discuss further here.

Anyhow, I open a new issue, delete this whole issue and go on with @sharnoff his last comment. And suggest not to comment there unless you really want to help and get this part improved. At least thanks for bringing it up.

https://github.com/crossterm-rs/crossterm/issues/422
